src/Microsoft.Health.Dicom.Core/Features/Query/QueryParser.cs

```diff
@@ -180,12 +213,18 @@ private bool TryParseDicomAttributeId(string attributeId, out DicomTag dicomTag)
             return succeed;
         }
 
-        private static void ValidateIfTagSupported(DicomTag dicomTag, string attributeId, QueryResource resourceType)
+        private static void ValidateIfTagSupported(DicomTag dicomTag, string attributeId, QueryResource resourceType, out CustomTagFilterDetails customTagFilterDetails, IDictionary<QueryResource, HashSet<CustomTagFilterDetails>> customTagMapping = null)
         {
+            customTagFilterDetails = null;
             HashSet<DicomTag> supportedQueryTags = QueryLimit.QueryResourceTypeToTagsMapping[resourceType];
 
             if (!supportedQueryTags.Contains(dicomTag))
             {
+                if (customTagMapping != null && customTagMapping.ContainsKey(resourceType) && customTagMapping[resourceType].TryGetValue(new CustomTagFilterDetails(dicomTag), out customTagFilterDetails))
```

Comment:
can you simply this? Maybe service only gets tags at a sepcific resourceType. Because a query will be at only one level.

Also, HashSet<DicomTag> supportedQueryTags = QueryLimit.QueryResourceTypeToTagsMapping[resourceType];
can we replaced by a function to get the list of both static and dynamic custom tahs